File: native/ios/Sources/WorkspaceModel.swift
import Amplify
import Combine
import Foundation
import SwiftUI

@MainActor
final class WorkspaceModel: ObservableObject {
    @Published private(set) var identity: Identity?
    @Published private(set) var appointments: [Appointment] = []
    @Published private(set) var next: String?
    @Published private(set) var busy = false
    @Published private(set) var message: String?
    @Published private(set) var challenge = false
    @Published private(set) var challengeInput: ChallengeInput = .code
    @Published private(set) var challengeChoices: [MFAType] = []
    @Published private(set) var visible = true
    let residency: String
    let content: MobileContent?
    let policies: MobilePolicies?
    let registration: AccountRegistration
    let cancellation: AppointmentCancellation
    let recovery: PasswordRecovery
    let profile: ProfileEnrollment
    private var profileObservation: AnyCancellable?
    private let auth: CognitoSession?
    private let api: AppointmentsAPI?
    private var operation: Task<Void, Never>?
    private var expiry: Task<Void, Never>?
    private var generation = 0
    var configured: Bool { auth != nil && api != nil }

    // A non-sensitive logout latch. Credentials remain managed by Amplify/Keychain.
    private var requiresExplicitSignIn: Bool {
        get { UserDefaults.standard.bool(forKey: "requires-explicit-sign-in") }
        set { UserDefaults.standard.set(newValue, forKey: "requires-explicit-sign-in") }
    }

    init() {
        residency = Bundle.main.object(forInfoDictionaryKey: "MediConnectResidency") as? String ?? ""
        content = try? MobileContent(data: BundledAssets.data("mobile-content"))
        policies = try? MobilePolicies(legal: BundledAssets.data("legal"), consent: BundledAssets.data("consent"))
        var profileService: ProfileAPI?
        var cancellationService: AppointmentCancellationAPI?
        var cancellationPolicy: CancellationContract?
        do {
            let config = try MobileConfiguration(data: BundledAssets.data("mobile-config"), residency: residency)
            let contract = try MobileContract(data: BundledAssets.data("mobile-contract"), policy: BundledAssets.data("session-policy"))
            let session = try CognitoSession(config: config, contract: contract)
            auth = session
            profileService = ProfileAPI(transport: NativeAPI(config: config), contract: contract, fetch: { try await session.fetch() })
            let appointments = AppointmentsAPI(config: config, contract: contract)
            api = appointments
            cancellationService = AppointmentCancellationAPI(appointments: appointments, transport: NativeAPI(config: config), contract: contract, fetch: { try await session.fetch() })
            cancellationPolicy = contract.cancellation
        } catch { auth = nil; api = nil }
        cancellation = AppointmentCancellation(service: cancellationService, policy: cancellationPolicy)
        profile = ProfileEnrollment(service: profileService, policyVersion: policies?.policyVersion ?? "")
        recovery = PasswordRecovery(service: auth)
        registration = AccountRegistration(service: policies == nil ? nil : auth)
        profileObservation = profile.$state.filter { $0.step == .ready }.sink { [weak self] value in
            Task { @MainActor [weak self] in
                guard let self, value.profile?.subject == identity?.subject, profile.state.step == .ready else { return }
                refresh()
            }
        }
    }

    private func clear(visible: Bool = true) {
        cancellation.close(clearSession: true)
        profile.close()
        generation += 1
        operation?.cancel()
        expiry?.cancel()
        identity = nil
        appointments = []
        next = nil
        busy = false
        challenge = false
        challengeInput = .code
        challengeChoices = []
        message = nil
        self.visible = visible
    }
    func hide() { registration.close(); recovery.close(); clear(visible: false) }
    func openRegistration() {
        guard configured, policies != nil else { return }
        requiresExplicitSignIn = true
        recovery.close(); clear(); registration.open()
    }
    func openRecovery() {
        registration.close()
        guard configured else { return }
        requiresExplicitSignIn = true
        clear()
        recovery.open()
    }
    func resume() {
        guard recovery.state.step == .closed, registration.state.step == .closed else { return }
        visible = true
        guard !busy, identity == nil, !requiresExplicitSignIn, let auth else { return }
        run(message: "sessionExpired") { [weak self] in
            let access = try await auth.fetch()
            try Task.checkCancellation()
            try await self?.accept(access)
        }
    }
    func signIn(email: String, password: String) {
        guard !busy, !email.isEmpty, !password.isEmpty, let auth else { return }
        requiresExplicitSignIn = true
        run(message: "signInFailed") { [weak self] in
            await auth.signOut()
            try Task.checkCancellation()
            let result = try await auth.signIn(email: email.trimmingCharacters(in: .whitespacesAndNewlines), password: password)
            try Task.checkCancellation()
            if case .confirmSignUp = result.nextStep { self?.registration.resumeConfirmation(email) }
            else { try await self?.complete(result) }
        }
    }
    func confirm(code: String) {
        guard !busy, challenge, let auth else { return }
        let selected = challengeChoices.isEmpty ? SignInChallenge.response(challengeInput, value: code)
            : SignInChallenge.selectionResponse(challengeChoices, value: code)
        guard let response = selected else { return }
        run(message: "signInFailed") { [weak self] in
            let result = try await auth.confirm(code: response)
            try Task.checkCancellation()
            try await self?.complete(result)
        }
    }
    private func complete(_ result: AuthSignInResult) async throws {
        guard let auth else { throw MobileFailure.configuration }
        if result.isSignedIn {
            let access = try await auth.fetch()
            try Task.checkCancellation()
            requiresExplicitSignIn = false
            try await accept(access)
        } else {
            let input = SignInChallenge.input(result.nextStep)
            challengeChoices = SignInChallenge.choices(result.nextStep)
            challenge = input != nil || !challengeChoices.isEmpty
            challengeInput = input ?? .code
            message = challenge ? nil : "additionalStep"
        }
    }
    private func accept(_ access: SessionAccess) async throws {
        identity = access.identity
        challenge = false
        challengeChoices = []
        expiry?.cancel()
        let expiresAt = access.identity.expiresAt
        expiry = Task { [weak self] in
            do { try await Task.sleep(until: .now + .seconds(max(0, expiresAt.timeIntervalSinceNow)), clock: .continuous) }
            catch { return }
            self?.clear()
            self?.message = "sessionExpired"
        }
        profile.open(access.identity)
    }
    func refresh(more: Bool = false) {
        guard !busy, profile.state.step == .ready, let identity, let auth, let api, !more || next != nil else { return }
        let cursor = more ? next : nil
        let previous = more ? appointments : []
        run(message: "unavailable") { [weak self] in
            let access = try await auth.fetch()
            try Task.checkCancellation()
            guard access.identity.subject == identity.subject, access.identity.role == identity.role else { throw MobileFailure.session }
            let page = try await api.load(access: access, cursor: cursor)
            try Task.checkCancellation()
            var ids = Set<String>()
            self?.appointments = (previous + page.items).filter { ids.insert($0.id).inserted }
            self?.next = page.next
        }
    }
    func openCancellation(_ appointment: Appointment) {
        guard !busy, profile.state.step == .ready, let identity, appointments.contains(where: { $0.id == appointment.id }) else { return }
        cancellation.open(identity, appointment: appointment)
    }
    func closeCancellation() {
        let confirmed = cancellation.state.step == .confirmed
        cancellation.close()
        if confirmed { refresh() }
    }
    func signOut() {
        registration.close()
        recovery.close()
        requiresExplicitSignIn = true
        clear()
        guard let auth else { return }
        run(message: "signInFailed") { await auth.signOut() }
    }
    private func run(message: String, block: @escaping @MainActor () async throws -> Void) {
        operation?.cancel()
        generation += 1
        let current = generation
        busy = true
        self.message = nil
        operation = Task { [weak self] in
            do { try await block() }
            catch is CancellationError { return }
            catch {
                guard let self, current == generation, !Task.isCancelled else { return }
                if (error as? HTTPFailure)?.status == 401 {
                    identity = nil
                    appointments = []
                    next = nil
                    self.message = "sessionExpired"
                } else if (error as? HTTPFailure)?.status == 403 { self.message = "accessDenied" }
                else { self.message = identity == nil ? message : "unavailable" }
            }
            if let self, current == generation { busy = false }
        }
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/WorkspaceModel.kt
package com.mediconnect.nativeapp

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.amplifyframework.auth.result.AuthSignInResult
import com.amplifyframework.auth.result.step.AuthSignInStep
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant

data class WorkspaceState(
    val identity: Identity? = null,
    val appointments: List<Appointment> = emptyList(),
    val next: String? = null,
    val busy: Boolean = false,
    val error: String? = null,
    val challenge: Boolean = false,
    val challengeInput: ChallengeInput = ChallengeInput.CODE,
    val visible: Boolean = true,
)

class WorkspaceModel(application: Application) : AndroidViewModel(application) {
    val content = (application as MediConnectApplication).content
    val policies = (application as MediConnectApplication).policies
    private val runtime = (application as MediConnectApplication).runtime.getOrNull()
    val configured = runtime != null
    val cancellation = AppointmentCancellation(runtime?.appointments, runtime?.contract?.cancellation, viewModelScope)
    val recovery = PasswordRecovery(runtime?.sessions, viewModelScope)
    val profile = ProfileEnrollment(runtime?.profiles, policies.policyVersion, viewModelScope)
    val registration = AccountRegistration(runtime?.sessions, viewModelScope)
    private val mutable = MutableStateFlow(WorkspaceState())
    val state = mutable.asStateFlow()
    private var operation: Job? = null
    private var expiry: Job? = null
    private var generation = 0L

    init {
        viewModelScope.launch {
            profile.state.collect { value ->
                val identity = mutable.value.identity
                if (value.step == ProfileStep.READY && identity != null && value.profile?.subject == identity.subject) refresh()
            }
        }
    }

    private fun reset(visible: Boolean = true) {
        cancellation.close(clearSession = true)
        profile.close()
        generation++
        operation?.cancel()
        expiry?.cancel()
        mutable.value = WorkspaceState(visible = visible)
    }

    fun hide() { registration.close(); recovery.close(); reset(visible = false) }

    fun openRegistration() {
        val runtime = runtime ?: return
        runtime.requiresExplicitSignIn = true
        recovery.close(); reset(); registration.open()
    }

    fun openRecovery() {
        val runtime = runtime ?: return
        runtime.requiresExplicitSignIn = true
        registration.close()
        reset()
        recovery.open()
    }

    fun resume() {
        if (recovery.state.value.step != RecoveryStep.CLOSED || registration.state.value.step != RegistrationStep.CLOSED) return
        if (mutable.value.visible && (mutable.value.identity != null || mutable.value.busy)) return
        mutable.value = mutable.value.copy(visible = true)
        val runtime = runtime ?: return
        if (runtime.requiresExplicitSignIn) return
        runOperation("sessionExpired") { accept(runtime.sessions.fetch().identity) }
    }

    fun signIn(email: String, password: String) {
        val runtime = runtime ?: return
        if (email.isBlank() || password.isEmpty() || mutable.value.busy) return
        runtime.requiresExplicitSignIn = true
        runOperation("signInFailed") {
            runtime.sessions.signOut()
            val result = runtime.sessions.signIn(email.trim(), password)
            if (result.nextStep.signInStep == AuthSignInStep.CONFIRM_SIGN_UP) registration.resumeConfirmation(email)
            else completeSignIn(result)
        }
    }

    fun confirm(code: String) {
        val runtime = runtime ?: return
        if (!mutable.value.challenge || code.isBlank() || mutable.value.busy) return
        val response = SignInChallenge.response(mutable.value.challengeInput, code) ?: return
        runOperation("signInFailed") { completeSignIn(runtime.sessions.confirm(response)) }
    }

    private suspend fun completeSignIn(result: AuthSignInResult) {
        val runtime = runtime ?: return
        if (result.isSignedIn) {
            val identity = runtime.sessions.fetch().identity
            runtime.requiresExplicitSignIn = false
            accept(identity)
        } else {
            val input = SignInChallenge.input(result.nextStep.signInStep)
            mutable.value = mutable.value.copy(challenge = input != null, challengeInput = input ?: ChallengeInput.CODE, error = if (input != null) null else "additionalStep")
        }
    }

    private suspend fun accept(identity: Identity) {
        val runtime = runtime ?: return
        mutable.value = WorkspaceState(identity = identity, busy = true)
        expiry?.cancel()
        expiry = viewModelScope.launch {
            delay(Duration.between(Instant.now(), identity.expiresAt).toMillis().coerceAtLeast(0))
            reset()
            mutable.value = mutable.value.copy(error = "sessionExpired")
        }
        profile.open(identity)

    }

    fun refresh(more: Boolean = false) {
        val runtime = runtime ?: return
        val current = mutable.value
        val identity = current.identity ?: return
        if (current.busy || profile.state.value.step != ProfileStep.READY || (more && current.next == null)) return
        runOperation("unavailable") {
            val page = runtime.appointments.load(identity, if (more) current.next else null)
            mutable.value = mutable.value.copy(
                appointments = (if (more) current.appointments + page.items else page.items).distinctBy { it.id },
                next = page.next,
            )
        }
    }

    fun openCancellation(appointment: Appointment) {
        val current = mutable.value
        val identity = current.identity ?: return
        if (current.busy || profile.state.value.step != ProfileStep.READY || current.appointments.none { it.id == appointment.id }) return
        cancellation.open(identity, appointment)
    }
    fun closeCancellation() {
        val confirmed = cancellation.state.value.step == CancellationStep.CONFIRMED
        cancellation.close()
        if (confirmed) refresh()
    }

    fun signOut() {
        registration.close()
        recovery.close()
        val runtime = runtime ?: return
        runtime.requiresExplicitSignIn = true
        reset()
        runOperation("signInFailed") { runtime.sessions.signOut() }
    }

    private fun runOperation(error: String, block: suspend () -> Unit) {
        operation?.cancel()
        val current = ++generation
        mutable.value = mutable.value.copy(busy = true, error = null)
        operation = viewModelScope.launch {
            try { block() }
            catch (cancelled: CancellationException) { throw cancelled }
            catch (failure: Exception) {
                if (current == generation) {
                    val authFailed = failure is ApiFailure && failure.status == 401
                    if (authFailed) mutable.value = WorkspaceState(error = "sessionExpired")
                    else mutable.value = mutable.value.copy(error = when {
                        failure is ApiFailure && failure.status == 403 -> "accessDenied"
                        mutable.value.identity != null -> "unavailable"
                        else -> error
                    })
                }
            } finally {
                if (current == generation) mutable.value = mutable.value.copy(busy = false)
            }
        }
    }
}
